File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict

def solution(ticket):
    answer = []
    ticket.sort(reverse = True)
    d = defaultdict(list)
    
    for s,e in ticket:
        d[s].append(e)

    st = ['ICN']
    path = []
    
    while st:
        now = st[-1]
        
        if not d[now] or len(d[now]) == 0:
            path.append(st.pop())
        else:
            st.append(d[now].pop())
    return path[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    static boolean[] visited;
    static String[] answer;
    static boolean found;
    
    public String[] solution(String[][] tickets) {
                
        visited = new boolean[tickets.length];
        answer = new String[tickets.length+1];
        found = false;
        
        // 도착지 기준 정렬
        Arrays.sort(tickets, (a,b) -> {
            if (!a[0].equals(b[0])) return a[0].compareTo(b[0]);
            return a[1].compareTo(b[1]);
        });

        
        answer[0] = "ICN";
        dfs("ICN", tickets, 0);
        return answer;
        
    }
    
    static void dfs(String now, String[][] tickets, int depth) {
        if (found) return;
        
        if (depth == tickets.length) {
            found = true;
            return;
        }
        
        for (int i=0;i<tickets.length;i++) {
            if (!visited[i] && now.equals(tickets[i][0])) {
                answer[depth+1] = tickets[i][1];
                visited[i] = true;
                dfs(tickets[i][1], tickets, depth+1);
                
                if (found) return;
                visited[i] = false;
            }
        }
    }
}